AzureMonitor: Fix custom namespaces

asimpson opened the pull request:
**What this PR does / why we need it**:
The request for listing metric namespaces can return three types of namespaces as documented [here]. We previously only handled the platform namespace `microsoft.insights/components`.

[here]: https://docs.microsoft.com/en-us/rest/api/monitor/metric-namespaces/list?tabs=HTTP#metricnamespacecollection

Related to #54826

**Which issue(s) this PR fixes**:
Fixes #54757
Fixes https://github.com/grafana/support-escalations/issues/3761

**Special notes for your reviewer**:
If you have access to Grafana's Azure dev account you can use the `AppInsightsTest` resource to test.
![image](https://user-images.githubusercontent.com/1048831/189265564-55026e14-c824-41e8-ac1e-a7fc22fb2291.png)

asimpson committed:
AzureMonitor: Properly handle custom namespaces

The request for listing metric namespaces can return three types of
namespaces as documented [here][doc]. We previously only handled the platform
namespace `microsoft.insights/components`.

[doc]: https://docs.microsoft.com/en-us/rest/api/monitor/metric-namespaces/list?tabs=HTTP#metricnamespacecollection

andresmgot commented on pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource.go:
```diff
@@ -113,8 +113,13 @@ func (e *AzureMonitorDatasource) buildQueries(queries []backend.DataQuery, dsInf
+		if azJSONModel.CustomNamespace != "" {
+			params.Add("metricnamespace", azJSONModel.CustomNamespace)
+		} else {
+			params.Add("metricnamespace", azJSONModel.MetricNamespace) // MetricName or MetricNames ?
```
the comment is now in the wrong  line but we likely don't need that comment at all

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNamespaceField.tsx:
```diff
@@ -43,7 +43,7 @@ const MetricNamespaceField: React.FC<MetricNamespaceFieldProps> = ({
       <Select
         inputId="azure-monitor-metrics-metric-namespace-field"
-        value={query.azureMonitor?.metricNamespace || null}
+        value={query.azureMonitor?.customNamespace || query.azureMonitor?.metricNamespace || null}
```
```suggestion
        value={value || null}
```
?

andresmgot reviewed (commented):
Thanks for this tricky fix! I have some minor comments.

Food for thought but a good UX improvement would be to not render the namespace selector if there are no "custom" namespaces (this is, namespaces apart from the one of the resource). In most of the cases we could skip that input. Not necessary to do that in this PR though.

asimpson committed:
fix: add test for buildQueries + customNamespace

asimpson committed:
fix: account for storageaccounts + customNamespace

asimpson committed:
fix: changes based on feedback

also updated tests to match new URL format which doesn't always populate
metricnames like it used to.

asimpson committed:
chore: add test for setQueryValue

Added this to wrap a test case around the storageaccount logic branch.

andresmgot commented on public/app/plugins/datasource/grafana-azure-monitor-datasource/types/query.ts:
```diff
@@ -47,6 +47,7 @@ export interface AzureMetricQuery {
   resourceName?: string;
   /** Resource type */
   metricNamespace?: string;
+  customNamespace?: string;
```
since we are differing from the naming in Azure Monitor API, it would be good to add some comment here explaining the difference between `metricNamespace` and `customNamespace` for future reference

asimpson replied:
Yeah, I can PR that this am.

andresmgot reviewed (approved):
LGTM! Thanks

andresmgot commented:
let me merge this. It seems that 9.1.5 release is on its way now just in case we miss that train :)
